src/RedisIPCMessenger.ts

```diff
@@ -0,0 +1,119 @@
+import Redis from 'ioredis';
+import IPCMessenger, {
+  Room,
+  Instance,
+  MessageCallback,
+  MessageTypes,
+  Message,
+} from './IPCMessenger';
+
+type ConstructorParams = {
```

Comment:
Do you insist to keep on using a params object instead of separate constructor parameters?

Reply:
I'm also for using params objects, even if it's a js concept.

Reply:
Why though? It literally provides no benefit - it's just extra lines of code.

Reply:
For the constructor params, yes, I prefer using named arguments. The whole object can be threated as a unit, a configuration, describing how the component should opperate. You also get the flexibility of adding new props between existing without worrying for the order in both the component constructing the client and the constructor itself.
Named arguments also provide a better explanation for the param, I personally find looking at something like `new RedisIPCMessenger('container-name', undefined, 1000, 500)`  very confusing, whay do those numbers even mean? Now I have to go through each of them and memorize which one is what. Plus everytime you want to make something optional you will have to reorder your whole list, or you end up with random undefineds in the middle of the list.

And for what benefit? Not having to write `{}`? Both approaches will have the same LOC when the list exceeds 120 chars or whatever and you have to put each of them on new line.

Reply:
> adding new props between existing without worrying

This is kind of a problem - you end up adding things not because they are necessary, but because it's easy, which encourages less thinking, which is never good.

> Named arguments also provide a better explanation for the param

Not in TypeScript - what I end up seeing as a user of this class is the following:

<img width="702" alt="Screen Shot 2021-02-11 at 12 30 49" src="https://user-images.githubusercontent.com/1444508/107625449-fbaed300-6c64-11eb-85e0-c41926f7a4a6.png">

There is a type, which I can't find anywhere (it's not exported). So I have to go into the source. Alternatively, I can **assume** that it's an object, and start the guessworks:

```
Argument of type '{}' is not assignable to parameter of type 'ConstructorParams'.
  Property 'instance' is missing in type '{}' but required in type 'ConstructorParams'.ts(2345)
```

So, the above tells me that there is an `instance` missing - but what type is that? Again, I must go into the source.

The situation doesn't improve if I import the type, so I still have to go into the source:

<img width="495" alt="Screen Shot 2021-02-11 at 12 47 16" src="https://user-images.githubusercontent.com/1444508/107627001-4cbfc680-6c67-11eb-8f99-a0d4f7873b9e.png">

I guess, after I've went into the source, I could have _memorized_ the arguments so it would be a better user experience next time.

The IDE situation is much better when using separate parameters - you see their names and their types immediatelly:

<img width="621" alt="Screen Shot 2021-02-11 at 12 32 31" src="https://user-images.githubusercontent.com/1444508/107625605-387aca00-6c65-11eb-87ab-032b4b968290.png">

> new RedisIPCMessenger('container-name', undefined, 1000, 500)

The problem with this piece of code is the use of magic numbers as constructor arguments. 

> Now I have to go through each of them and memorize which one is what.

If you only hover in your IDE, it will show what is what immediately - no need to memorize anything (as opposed to the constructor object, where memorization is the only option):

<img width="629" alt="Screen Shot 2021-02-11 at 12 35 35" src="https://user-images.githubusercontent.com/1444508/107625882-a6bf8c80-6c65-11eb-822e-62d9816cbf9a.png">

> Plus everytime you want to make something optional you will have to reorder your whole list, or you end up with random undefineds in the middle of the list.

Yes, this is a valid argument - you will have to do that. However, arguably, it's better practice to not have optional parameters in the constructor altogether.

> Both approaches will have the same LOC

Well, yes, more or less, you're right. However, arguably, it's a better practice to have a few dependencies as possible  (and, constructor arguments are dependencies), which would make for fewer situations, where you need to go multi-line. With the explicit type you have to go multi-line every time.

Overall, I don't see many actual every-day benefits to the approach. Both the optionals argument and the multi-line constructor arguments can be argued one way or another.

However, the downside of not being able to use your IDE to help you is tangible and will slow you down - until you memorize the arguments and their types.

Reply:
> This is kind of a problem - you end up adding things not because they are necessary, but because it's easy, which encourages less thinking, which is never good.

On the other hand, making reordering more difficult discourages people from taking the time to improve the order of the arguments.

> Not in TypeScript - what I end up seeing as a user of this class is the following:

Yes, but only after hovering cursor over the value.

> So, the above tells me that there is an instance missing - but what type is that? Again, I must go into the source.

You can just trigger the intellisense (`ctrl + space`) and you will see.

![image](https://user-images.githubusercontent.com/13086078/107629248-81814d00-6c6a-11eb-918b-e6a322dd64ca.png)

> If you only hover in your IDE, it will show what is what immediately - no need to memorize anything

For a greater number of arguments, you will have to hover each of them and memorize them in case there is a relation between them, e.g. we defined 3+ intervals as magic numbers.

>  (as opposed to the constructor object, where memorization is the only option):

Why memorize the constructor if each value is accompanied by a key?

> few dependencies as possible (and, constructor arguments are dependencies),

Yes, that's why I prefer threating them as a single variable. Constructors are functions, and as such, I don't see why we would want to allow great number of args.

Reply:
> On the other hand, making reordering more difficult discourages people from taking the time to improve the order of the arguments.

Why are you using this as an argument? With a DI container, which, when used correctly, calls the constructor for you at instantiation, you only have to change the implementation of the constructor internally and you're set. At worst, you also have to change your factory, but the TS compiler will help you with that too. If  you're calling `new MyClass` at 1000 places, we have a bigger problem.

> You can just trigger the intellisense (ctrl + space) and you will see.

True, but I have to trigger the suggestion for every argument manually, which is annoying.

> Yes, but only after hovering cursor over the value.

No, when I type `new Something(` I already see it pop up automatically.

> For a greater number of arguments, you will have to hover each of them and memorize them in case there is a relation between them, e.g. we defined 3+ intervals as magic numbers.

I don't understand - could you please elaborate? First, you shouldn't use magic numbers/strings in general. Second, you hover over the constructor and it shows all the params with their types - no need to manually trigger suggest for  every key.

> Yes, that's why I prefer threating them as a single variable. Constructors are functions, and as such, I don't see why we would want to allow great number of args.

But it's really a delusion that you only have one argument - even the type of the argument is called `params`, plural.

Reply:
> With a DI container, which, when used correctly, calls the constructor for you at instantiation, you only have to change the implementation of the constructor internally and you're set. 

We are not using a TS DI container, the last time we had such a discussion we decided not on using one because the only one we found was using decorators.

> No, when I type new Something( I already see it pop up automatically.

We are not always typing, most of the time people are just reading the code.

> I don't understand - could you please elaborate? First, you shouldn't use magic numbers/strings in general. Second, you hover over the constructor and it shows all the params with their types - no need to manually trigger suggest for every key.

By using a list you will end up with magic numbers and strings, unless you hover. In the case of greater number of such params, you will have to memorize the value for each one only after hover.

> But it's really a delusion that you only have one argument - even the type of the argument is called params, plural.

Technically yes, although we threat them more as a config which we consider to be complete unit. They do include uncorrelated props, so maybe this assumption is wrong to begin with.

Reply:
In TS projects we will still use the current DI container, which works perfectly fine with TS or without.

You must not use magic numbers at all, reading or writing. Why would you?

Let's further discuss with the rest of the team.